### PR TITLE
2.8 - Populate Link-Layer Device Origin from Instance Poller

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

This patch refactors the instance-poller API server logic for updating link-layer devices, so that device address origin is managed.

Observed devices and addresses have provider ID set as appropriate, observed addresses have their origin set to _provider_ and unrecognised addresses have origin set to _machine_ indicating that the provider relinquishes control of them.

The opportunity is taken here to reduce reliance on the unwieldy methods `SetParentLinkLayerDevicesBeforeTheirChildren` and `SetDevicesAddressesIdempotently`. Instead a modular `ModelOperation` approach is preferred. The hope is that these methods can be removed entirely at some point.

In the interests of keeping patch size reviewable, there are the following notable ommissions:
- We do not currently set provider subnet ID or provider network ID for addresses. It was noticed that although network ID is set in args, there is no field for it in state. This will be addressed in a follow-up patch. Omitting the subnet ID while we work to include it with the network ID is OK, since it does not break any existing functionality.
- Controller machines did not previously, and still do not handle devices correctly. This is because the machiner updates provider link-layer data via the networking common API for controllers. This results in duplicate devices on AWS, _unsupported0_ from the provider and _eth0_ from the machine, with only the provider device having a provider ID. This will be rectified in a follow-up.

## QA steps

- Bootstrap to AWS and add a machine.
- Check that the machine's main NIC (ens3 for example) gets a provider ID in the `linklayerdevices` Mongo collection.
- Check that the address for the NIC gets origin `provider` in the `ip.addresses` collection.
- Same steps as above on MAAS, but the address should also get a provider ID.

## Documentation changes

None.

## Bug reference

- Advances https://bugs.launchpad.net/bugs/1733968
- https://bugs.launchpad.net/bugs/1855263

### CODE REVIEW

File: apiserver/facades/controller/instancepoller/mock_test.go [hunk 1]
```diff
@@ -423,47 +411,13 @@ func (m *mockMachine) Status() (status.StatusInfo, error) {
...
-func (m mockLinkLayerDevice) ParentName() string                { return m.parentName }
-func (m mockLinkLayerDevice) Addresses() ([]instancepoller.StateLinkLayerDeviceAddress, error) {
-	return m.addresses, nil
+// AssertAliveOp implements StateMachine.
+func (m *mockMachine) AssertAliveOp() txn.Op {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.MethodCall(m, "Status")
+	return txn.Op{C: "machine-alive"}
```
Comment (manadart) [inline apiserver/facades/controller/instancepoller/mock_test.go:420]: Being a little creative with ops, so we can interrogate which were generated and in which order.

File: apiserver/facades/controller/instancepoller/mock_test.go [hunk 2]
```diff
@@ -258,15 +250,29 @@ func (m *mockState) SetSpaceInfo(infos network.SpaceInfos) {
 // used with watcher helpers/checkers.
 func (m *mockState) StartSync() {}
 
+func (m *mockState) ApplyOperation(op state.ModelOperation) error {
+	m.MethodCall(m, "ApplyOperation", op)
+
+	ops, err := op.Build(0)
+	if err == nil {
+		// Register the generated transactions so that we can interrogate them.
+		m.MethodCall(m, "ApplyOperation.Build", ops)
```
Comment (manadart) [inline apiserver/facades/controller/instancepoller/mock_test.go:259]: Then this sub-call can be used to retrieve them.

**Comment by manadart** [comment]
!!build!!

File: state/linklayerdevices_ipaddresses.go [hunk 3]
```diff
@@ -227,6 +226,72 @@ func (addr *Address) Remove() (err error) {
 	return addr.st.db().RunTransaction(ops)
 }
 
+// SetOriginOps returns the transaction operations required to set the input
```
Comment (achilleasa) [inline state/linklayerdevices_ipaddresses.go:229]: Reviewed till here; will resume on Monday

File: state/linklayerdevices.go [hunk 4]
```diff
@@ -191,6 +191,35 @@ func (dev *LinkLayerDevice) parentDocID() string {
...
+// provider ID for the link-layer device.
+func (dev *LinkLayerDevice) SetProviderIDOps(id network.Id) ([]txn.Op, error) {
+	// We only set the provider ID if it was previously empty.
+	if dev.doc.ProviderID != "" || id == "" || dev.doc.ProviderID == id.String() {
+		return nil, nil
+	}
+
+	// Since we assume that we are now setting the ID for the first time,
+	// ensure that it has not already been used to identify another device.
+	exists, err := dev.st.networkEntityGlobalKeyExists("linklayerdevice", id)
```
Comment (achilleasa) [inline state/linklayerdevices.go:204]: This method makes a DB query which runs outside the generated txn ops below. Can we make it return a txn.Op instead to completely eliminate the possibility of a dup ID entry?

File: state/linklayerdevices_ipaddresses.go [hunk 5]
```diff
@@ -227,6 +226,72 @@ func (addr *Address) Remove() (err error) {
...
+func (addr *Address) SetOriginOps(origin network.Origin) []txn.Op {
+	if addr.Origin() == origin {
+		return nil
+	}
+
+	updates := bson.D{bson.DocElem{Name: "$set", Value: bson.M{"origin": origin}}}
+
+	removeProviderID := origin == network.OriginMachine &&
+		addr.Origin() == network.OriginProvider &&
+		addr.ProviderID() != ""
```
Comment (achilleasa) [inline state/linklayerdevices_ipaddresses.go:242]: What's the purpose of the check on L242? If the provider ID was not populated correctly for some reason but we relinquish control to the machiner shouldn't it assume control of the address regardless of whether the provider ID is set or not?

File: state/linklayerdevices_ipaddresses.go [hunk 6]
```diff
@@ -227,6 +226,72 @@ func (addr *Address) Remove() (err error) {
...
+// Setting the provider ID updates the address origin to provider.
+func (addr *Address) SetProviderIDOps(id network.Id) ([]txn.Op, error) {
+	// We only set the provider ID if it was previously empty.
+	if addr.doc.ProviderID != "" || id == "" || addr.doc.ProviderID == id.String() {
+		return nil, nil
+	}
+
+	// Since we assume that we are now setting the ID for the first time,
+	// ensure that it has not already been used to identify another device.
+	exists, err := addr.st.networkEntityGlobalKeyExists("address", id)
```
Comment (achilleasa) [inline state/linklayerdevices_ipaddresses.go:272]: See my comment above (linklayerdevices) for making this part of the txn.

File: state/linklayerdevices.go (hunk 4 above)
Comment (manadart) [reply state/linklayerdevices.go:204]: `networkEntityGlobalKeyOp` used to insert the ID below already asserts that the doc is missing.

Remember that this method is being called from with the model operation `Build` command, so we will error at this point with an appropriate message if a dup is dedected, but the assertion still handles a race.

File: state/linklayerdevices_ipaddresses.go (hunk 5 above)
Comment (manadart) [reply state/linklayerdevices_ipaddresses.go:242]: If the origin is changing, we will *always* run at least the update generated on line 238 - setting the origin.

If it happens to be changing back to "machine" and there is a provider ID set, we also update the doc to remove the provider ID *and* we remove the provider ID from the global collection.

**Comment by manadart** [comment]
$$merge$$

**Comment by manadart** [comment]
$$merge$$

**Comment by manadart** [comment]
Mongo connection failure. Couldn't replicate locally.
```
10:35:12 [LOG] 0:02.001 DEBUG juju.provider.dummy resetting MgoServer
10:35:12 [LOG] 0:02.002 DEBUG juju.testing net.Dial(127.0.0.1:39137) failed with dial tcp 127.0.0.1:39137: connect: connection refused
```
$$merge$$

**Pull request merged**